solidity/contracts/ValidatorRegistration.sol

```diff
@@ -1,23 +1,113 @@
+    uint public constant DEPOSIT_SIZE = 32 ether;
+    uint public constant DEPOSITS_FOR_CHAIN_START = 8;
+    uint public constant MIN_TOPUP_SIZE = 1 ether;
+    uint public constant GWEI_PER_ETH = 10 ** 9;
+    // Setting MERKLE_TREE_DEPTH to 16 instead of 32 due to gas limit
+    uint public constant MERKLE_TREE_DEPTH = 16;
+    uint public constant SECONDS_PER_DAY = 86400;
 
-      used_pubkeys[_pubkey] = true;
+    mapping (uint => bytes) public receiptTree;
+    uint public totalDepositCount;
 
-      emit Deposit(_pubkey, _withdrawal_shard_id, _withdrawal_address, _randao_commitment);
-  }
+    // When users wish to become a validator by moving ETH from
+    // 1.0 chain to the 2.0 chain, they should call this function
+    // sending along DEPOSIT_SIZE ETH and providing depositParams
+    // as a simple serialize'd DepositParams object of the following
+    // form: 
+    // {
+    //    'pubkey': 'int256',
+    //    'proof_of_possession': ['int256'],
+    //    'withdrawal_credentials`: 'hash32',
+    //    'randao_commitment`: 'hash32'
+    // }
+    function deposit(
+        bytes depositParams
+    )
+        public
+        payable
+    {
+        uint index = totalDepositCount + 2 ** MERKLE_TREE_DEPTH;
+        bytes memory msgGweiInBytes = toBytes(msg.value);
+        bytes memory timeStampInBytes = toBytes(block.timestamp);
+        bytes memory depositData = mergeBytes(mergeBytes(msgGweiInBytes, timeStampInBytes), depositParams);
 
-}
+        emit Eth1Deposit(receiptTree[1], depositParams, totalDepositCount);
+
+        receiptTree[index] = abi.encodePacked(keccak256(depositData));
+        for (uint i = 0; i < MERKLE_TREE_DEPTH; i++) {
+            index = index / 2;
+            receiptTree[index] = abi.encodePacked(keccak256(mergeBytes(receiptTree[index * 2], receiptTree[index * 2 + 1])));
+        }
+
+        require(
+            msg.value <= DEPOSIT_SIZE,
+            "Deposit can't be greater than DEPOSIT_SIZE."
+        );
+        require(
+            msg.value >= MIN_TOPUP_SIZE,
+            "Deposit can't be lesser than MIN_TOPUP_SIZE."
+        );
+        if (msg.value == DEPOSIT_SIZE) {
+            totalDepositCount++;
+        }
+
+        // When ChainStart log publishes, beacon chain node initializes the chain and use timestampDayBoundry
+        // as genesis time.
+        if (totalDepositCount == DEPOSITS_FOR_CHAIN_START) {
+            uint timestampDayBoundry = block.timestamp - block.timestamp % SECONDS_PER_DAY + SECONDS_PER_DAY;
+            bytes memory timestampDayBoundryBytes = toBytes(timestampDayBoundry);
+            emit ChainStart(receiptTree[1], timestampDayBoundryBytes);
+        }
+    }
+
+    function getReceiptRoot() public constant returns (bytes) {
+        return receiptTree[1];
+    }
+
+    function toBytes(uint x) private constant returns (bytes memory) {
+        bytes memory b = new bytes(32);
+        assembly {
+          mstore(add(b, 32), x)
```

Comment:
does this work? I thought the first 32 bytes of a `bytes` type was the length, but this looks like it just stores x at location b+32

Reply:
also I'm not sure why we are using `bytes` and not `bytes32` here if the returned byte size is always 32

Reply:
I think this depends on the behavior of `concat` again. This function prepares values for concatenation, and if their lengths are included after concatenation then we'll need lengths here. Otherwise, a switch to `bytes32` is needed as you said.

Reply:
After checking [the spec](https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#vyper-code) again, it looks like we can have this function restrict its return value length to 8 bytes as needed. Again, assuming `concat` doesn't include argument lengths.

Reply:
As it appears in [the unit tests](https://github.com/ethereum/vyper/blob/1472f3f41aa510bf510471695a023f2cf8bdf385/tests/parser/functions/test_concat.py), concat doesn't include the 32 bytes in `bytes` types reserved for length when concatenating. With this being said, a `bytes32` return type in this function wouldn't suffice, as all 32 bytes would be packed by `encodePacked`. We should keep the `bytes` type but restrict the length two 8 bytes instead of 32.